src/util/lp/rooted_mons.h

```diff
@@ -0,0 +1,239 @@
+    std::unordered_map<unsigned, index_with_sign>                    m_reg_to_rm;
+
+    void print_stats(std::ostream& out) const {
+        static double ratio = 1;
+        double s = 0;
+        for (const auto& p : m_map) {
+            s += m_vector[p.second].m_mons.size();
+        }
+        double r = s /m_map.size();
+        if (r > ratio) {
+            ratio = r;
+            out << "rooted mons "  << m_map.size() << ", ratio = " << r << "\n";
+        }
+    }
+
+    const svector<unsigned>& to_refine() { return m_to_refine; }
+
+    bool list_is_consistent(const vector<index_with_sign>& list, const std::unordered_set<unsigned>& to_refine_reg) const {
+        bool t = to_refine_reg.find(list.begin()->index()) == to_refine_reg.end();
+        for (const auto& i_s : list) {
+            bool tt = to_refine_reg.find(i_s.index()) == to_refine_reg.end();
+            if (tt != t) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    bool list_contains_to_refine_reg(const vector<index_with_sign>& list, const std::unordered_set<unsigned>& to_refine_reg) const {
+        // the call should happen when no derived sign lemma is found, so the assert below should hold
+        SASSERT(list_is_consistent(list, to_refine_reg));
+        return !(to_refine_reg.find(list.begin()->index()) == to_refine_reg.end());
+    }
+    
+    void init_to_refine(const std::unordered_set<unsigned>& to_refine_reg) {
+        SASSERT(m_to_refine.empty());
+        for (unsigned i = 0; i < vec().size(); i++) {
+            if (list_contains_to_refine_reg(vec()[i].m_mons, to_refine_reg))
+                m_to_refine.push_back(i);
+        }
+        TRACE("nla_solver", tout << "rooted m_to_refine =["; print_vector(m_to_refine, tout) << "]\n";);
+    }
+    
+    void clear() {
+        m_map.clear();
+        m_vector.clear();
+        m_mons_containing_var.clear();
+        m_proper_factors.clear();
+        m_to_refine.clear();
+        m_reg_to_rm.clear();
+    }
+    
+    const vector<rooted_mon>& vec() const { return m_vector; }
+    vector<rooted_mon>& vec() { return m_vector; }
+
+    const std::unordered_map<svector<lpvar>, unsigned, hash_svector> & map() const {
+        return m_map;
+    }
+
+    std::unordered_map<svector<lpvar>, unsigned, hash_svector> & map() {
+        return m_map;
+    }
+
+    const std::unordered_map<lpvar, std::unordered_set<unsigned>>& var_map() const {
+        return m_mons_containing_var;
+    }
+
+    std::unordered_map<lpvar, std::unordered_set<unsigned>>&  var_map() {
+        return m_mons_containing_var;
+    }
+
+    std::unordered_map<unsigned, std::unordered_set<unsigned>>& proper_factors() {
+        return m_proper_factors;
+    }
+
+    const std::unordered_map<unsigned, std::unordered_set<unsigned>>& proper_factors() const {
+        return m_proper_factors;
+    }
+
+    void reduce_set_by_checking_proper_containment(std::unordered_set<unsigned>& p,
```

Comment:
elsewhere, we tend to use pattern:

auto j = p.begin();
for (auto i = p.begin(); i != p.end(); ++i) 
   if (test(i)) *j++ =*i;
p.set_end(j);